### PR TITLE
added categories

### PR DESCRIPTION
**Pull request by DmytroFedoryshyn**

### CODE REVIEW

File: src/main/java/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    @Operation(summary = "Create a new category")
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    @ApiResponse(responseCode = "201", description = "Category created", content = {
+            @Content(mediaType = "application/json", schema =
+            @Schema(implementation = CategoryResponseDto.class),
+                    examples = @ExampleObject(value = "{\"id\": 1, \"name\":"
+                            + " \"Fiction\", \"description\": \"Fictional books\"}"))
+    })
+    public CategoryResponseDto createCategory(@Valid @RequestBody
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/CategoryController.java:54]: PreAuthorize?

File: src/main/java/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    @Operation(summary = "Update a category by ID")
+    @PutMapping("/{id}")
+    @ApiResponse(responseCode = "200", description = "Category updated", content = {
+            @Content(mediaType = "application/json", schema =
+            @Schema(implementation = CategoryResponseDto.class),
+                    examples = @ExampleObject(value = "{\"id\": 1, \"name\":"
+                            + "\"Fiction\", \"description\": \"Updated description\"}"))
+    })
+    @ApiResponse(responseCode = "404", description = "Category not found")
+    public CategoryResponseDto updateCategory(@PathVariable Long id,
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/CategoryController.java:103]: PreAuthorize?

File: src/main/java/bookstore/mapper/BookMapper.java [hunk 3]
```diff
@@ -1,16 +1,64 @@
...
+    BookResponseDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
+
+    @Named("mapLongsToCategories")
+    default Set<Category> mapLongsToCategories(Set<Long> categoryIds) {
+        return categoryIds.stream()
+                .map(id -> {
+                    Category category = new Category();
+                    category.setId(id);
+                    return category;
+                })
```
Comment (Rommelua) [inline src/main/java/bookstore/mapper/BookMapper.java]:
```suggestion
                .map(Category ::new)
```

File: src/main/java/bookstore/service/CategoryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    public CategoryResponseDto save(CreateCategoryRequestDto categoryDto) {
+        return mapper.toDto(repository.save(mapper.toEntity(categoryDto)));
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, CreateCategoryRequestDto categoryDto) {
+        Category category = mapper.toEntity(categoryDto);
```
Comment (Rommelua) [inline src/main/java/bookstore/service/CategoryServiceImpl.java]:
```suggestion
    public CategoryResponseDto update(Long id, CreateCategoryRequestDto categoryDto) {
            Category dbCategory = categoryRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Category not found with ID: " + id));
        Category category = mapper.toEntity(categoryDto);
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "categories")
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = false")
+@Setter
+@Getter
+@RequiredArgsConstructor
```
Comment (aaananas) [inline src/main/java/bookstore/model/Category.java]:
```suggestion
@NoArgsConstructor
```
RequiredArgsConstructor creates a constructor with final fields. In your case you should use NoArgsConstructor

File: src/main/java/bookstore/repository/book/BookRepository.java [hunk 6]
```diff
@@ -1,10 +1,15 @@
...
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @Query("select distinct b from Book b join b.categories c where c.id = :id")
+    List<Book> findAllByCategories_Id(Long id, Pageable pageable);
```
Comment (aaananas) [inline src/main/java/bookstore/repository/book/BookRepository.java]:
```suggestion
    List<Book> findAllByCategoryId(Long id, Pageable pageable);
```

File: src/main/java/bookstore/service/BookServiceImpl.java [hunk 7]
```diff
@@ -68,4 +69,13 @@ public List<BookResponseDto> search(BookSearchParametersDto searchParametersDto,
                 .stream()
                 .map(bookMapper::toBookDto).toList();
     }
+
+    @Override
+    public List<BookResponseDtoWithoutCategoryIds> findAllByCategories_Id(
```
Comment (aaananas) [inline src/main/java/bookstore/service/BookServiceImpl.java]:
```suggestion
    public List<BookResponseDtoWithoutCategoryIds> findAllByCategoryId(
```

File: src/main/java/bookstore/dto/book/BookResponseDtoWithoutCategoryIds.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package bookstore.dto.book;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class BookResponseDtoWithoutCategoryIds {
```
Comment (aaananas) [inline src/main/java/bookstore/dto/book/BookResponseDtoWithoutCategoryIds.java]:
```suggestion
public class BaseBookResponseDto {
```

**Review by aaananas — CHANGES_REQUESTED** [review]

**Review by Rommelua — APPROVED** [review]

**Pull request merged**